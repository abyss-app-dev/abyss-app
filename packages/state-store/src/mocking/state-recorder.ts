import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
