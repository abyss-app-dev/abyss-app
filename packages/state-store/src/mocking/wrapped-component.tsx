import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
