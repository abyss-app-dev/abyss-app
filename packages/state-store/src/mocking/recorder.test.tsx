import { render } from '@testing-library/react';
import { useEffect, useState } from 'react';
import { describe, expect, test } from 'vitest';
import { StateRecorder } from './state-recorder';
import { TestRecorder } from './wrapped-component';

describe('recorder', () => {
    test('demonstrates the testing pattern using recorder', async () => {
        const recorder = new StateRecorder();
        render(<TestRecorder recorder={recorder} functionToTest={() => 1} />);
        expect(recorder.getStates()).matchSnapshot();
    });

    test('demonstrates the testing pattern using recorder with state updates', async () => {
        const recorder = new StateRecorder();

        const test = () => {
            const [state, setState] = useState(0);

            useEffect(() => {
                if (state < 3) {
                    setState(state + 1);
                }
            }, [state]);

            return state;
        };

        render(<TestRecorder recorder={recorder} functionToTest={test} />);
        expect(recorder.getStates()).matchSnapshot();
    });
});
