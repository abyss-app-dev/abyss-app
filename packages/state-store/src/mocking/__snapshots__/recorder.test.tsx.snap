// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`recorder > demonstrates the testing pattern using recorder 1`] = `
[
  1,
]
`;

exports[`recorder > demonstrates the testing pattern using recorder with state updates 1`] = `
[
  0,
  1,
  2,
  3,
]
`;
